cdap-ui/app/cdap/components/PluginJSONCreator/Create/Content/WidgetCollection/WidgetAttributesCollection/WidgetAttributeInput/AttributeKeyValueInput/index.tsx

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import * as React from 'react';
+import WidgetWrapper from 'components/ConfigurationGroup/WidgetWrapper';
+
+const AttributeKeyValueInput = ({
+  keyField,
+  valueField,
+  keyRequired,
+  valueRequired,
+  widgetToAttributes,
+  setWidgetToAttributes,
+  widgetID,
+  field,
+}) => {
+  const onKeyChange = (newVal) => {
+    setWidgetToAttributes((prevObjs) => ({
+      ...prevObjs,
+      [widgetID]: {
+        ...prevObjs[widgetID],
+        [field]: {
+          ...prevObjs[widgetID][field],
+          [keyField]: newVal,
+        },
+      },
+    }));
+  };
+
+  const onValueChange = (newVal) => {
+    setWidgetToAttributes((prevObjs) => ({
+      ...prevObjs,
+      [widgetID]: {
+        ...prevObjs[widgetID],
+        [field]: {
+          ...prevObjs[widgetID][field],
+          [valueField]: newVal,
+        },
+      },
+    }));
+  };
+
+  const keyWidget = {
+    label: field + ' ' + keyField,
+    name: keyField,
+    'widget-type': 'textbox',
+    'widget-attributes': {},
+  };
+
+  const valueWidget = {
+    label: field + ' ' + valueField,
+    name: keyField,
+    'widget-type': 'textbox',
+    'widget-attributes': {},
+  };
+
+  const keyProperty = {
+    required: keyRequired,
+    name: keyField,
+  };
+
+  const valueProperty = {
+    required: valueRequired,
+    name: valueField,
+  };
+
+  const currentKey = widgetToAttributes[widgetID][field][keyField]
```

Comment:
use `objectQuery` for this, it will automatically set things to null if the path doesn't exist. It's also safer. `services/helpers`

Reply:
since we change this code in the later PR to use immutable JS (https://github.com/cdapio/cdap/blob/12a3974e99884ac7ba9a8829ed665243a241df77/cdap-ui/app/cdap/components/PluginJSONCreator/Create/Content/WidgetCollection/WidgetAttributesCollection/WidgetAttributeInput/AttributeKeyValueInput/index.tsx#L63), we decided to leave this for now.

However, I will add a safeguard to make sure those value exists before calling a `.get` through `Immutable.js`.